jl4-lsp/src/LSP/L4/Actions.hs

```diff
@@ -96,6 +103,54 @@ gotoDefinition pos m positionMapping = do
   newRange <- toCurrentRange positionMapping lspRange
   pure (Location (fromNormalizedUri range.moduleUri) newRange)
 
+
+-- ----------------------------------------------------------------------------
+-- Ladder evalApp
+-- ----------------------------------------------------------------------------
+
+evalApp
+  :: MonadIO m
+  => TypeCheckResult
+  -> Ladder.EvalAppRequestParams
+  -> RecentlyVisualised
+  -> EL.Environment
+  -> ExceptT (TResponseError method) m Aeson.Value
+evalApp tcRes evalParams recentViz evalEnv =
+  case Ladder.lookupEvalAppMaker recentViz.vizState evalParams.appExpr of
+    Nothing -> defaultResponseError "No eval app directive maker found" -- TODO: Improve error codehere
+    Just evalAppMaker -> do
+      let evalAppDirective = evalAppMaker evalParams
+      (_, results) <- liftIO $ EL.execEvalModuleWithEnv evalEnv (evalAppDirective `prependToModule` tcRes.module')
+      Aeson.toJSON <$> getEvalResult results
+  where
+    prependToModule :: TopDecl Resolved -> Module Resolved -> Module Resolved
+    prependToModule newDecl (MkModule ann nuri (MkSection sann sresolved maka decls)) =
+      MkModule ann nuri (MkSection sann sresolved maka (newDecl : decls))
+
+    evalResultToLadderEvalAppResult :: Monad m => EL.EvalDirectiveResult -> ExceptT (TResponseError method) m EvalAppResult
+    evalResultToLadderEvalAppResult (EL.MkEvalDirectiveResult _ res) = case res of
+      Right (EL.MkNF val) -> 
+        case val of
+          EL.ValConstructor r [] -> EvalAppResult <$> toUBoolValue r
+          _                      -> throwExpectBoolResultError
+      Right EL.ToDeep    -> throwExpectBoolResultError
```

Comment:
I don't think that all of these should have the same error, especially the ToDeep one should not be "impossible", should it?

Reply:
I changed it to `defaultResponseError "Evaluation exceeded maximum depth limit"`, but let me know if you think of a better error message!